### PR TITLE
Use std::vector instead of sorted_array, remove base/tl/sorted_array.h

### PR DESCRIPTION
**Pull request by Robyt3**
This replaces all usages of `sorted_array` with `std::vector`. This requires the following general changes:

- `add_unsorted` is replaced with `push_back`.
- `add` is replaced with `push_back` and subsequent `sort` or `stable_sort` must be ensured.
  - In some cases, immediately sorting the entire list after adding an item was unavoidable. Previously items were added at the correct position, which was O(N) because all items after the inserted one had to be moved in any case.
- `sort_range` is replaced with `sort` or `stable_sort`.
- `size` returns a `size_t` instead of `int`, so to fix sign comparison warnings, casts are added where necessary or types of loop variables are changed to `size_t` where possible. For-each loops are also used where possible / where required by clang-tidy.
- `find_binary` is replaced with `std::equal_range`. This can only find items of the same type, so some wrappers, that only have the relevant fields set, need to be created for searching.

In terms of behavior, this should not change anything, except maybe `CLocalizationDatabase` for the better. As far as I understand it, at lot of the code there was not doing anything. It assumes that binary search can return a range of multiple entries, but the equality/comparison function is based on hash and context hash. This means that any item in this range will match the given hash and context hash already, so all of the following checks are redundant. I changed this to first do a lookup with the hash and context hash and if that fails do another lookup with the default context hash.

I have also already replaced `base/tl/array.h` with `std::vector`, removing all of `base/tl` except `threading.h`. I'll make a separate PR later because this caused a lot more changes especially in the editor that I first want to test and review myself.

Naming of `array`/`sorted_array`/`std::vector` variables was rather inconsistent (sometimes prefix `a` or `l` is used), so I chose to not use any prefix for all new `std::vector`s.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/localization.cpp [hunk 1]
```diff
@@ -42,7 +41,7 @@ void CLocalizationDatabase::AddString(const char *pOrgStr, const char *pNewStr,
 	s.m_Hash = str_quickhash(pOrgStr);
 	s.m_ContextHash = str_quickhash(pContext);
 	s.m_pReplacement = m_StringsHeap.StoreString(*pNewStr ? pNewStr : pOrgStr);
-	m_Strings.add(s);
+	m_Strings.push_back(s);
```
Comment (Learath2) [inline src/game/localization.cpp:44]: This should probably be:
```suggestion
	m_Strings.emplace_back(CString(str_quickhash(pOrgStr), str_quickhash(pContext), m_StringsHeap.StoreString(*pNewStr ? pNewStr : pOrgStr));
```

Avoiding an extra copy

File: src/game/client/components/skins.cpp [hunk 2]
```diff
@@ -415,20 +419,23 @@ int CSkins::FindImpl(const char *pName)
 	if(str_find(pName, "/") != 0)
 		return -1;
 
-	auto d = ::find_binary(m_aDownloadSkins.all(), pName);
-	if(!d.empty())
+	CDownloadSkin DownloadNeedle;
+	mem_zero(&DownloadNeedle, sizeof(DownloadNeedle));
+	str_copy(DownloadNeedle.m_aName, pName, sizeof(DownloadNeedle.m_aName));
+	auto DownloadRange = std::equal_range(m_DownloadSkins.begin(), m_DownloadSkins.end(), DownloadNeedle);
```
Comment (Learath2) [inline src/game/client/components/skins.cpp:425]:
```suggestion
	auto &[RangeBegin, RangeEnd] = std::equal_range(m_DownloadSkins.begin(), m_DownloadSkins.end(), DownloadNeedle);
```

Could be more readable than `.first`, `.second`

File: src/game/client/components/skins.cpp [hunk 3]
```diff
@@ -295,7 +295,8 @@ int CSkins::LoadSkin(const char *pName, CImageInfo &Info)
 		Console()->Print(IConsole::OUTPUT_LEVEL_ADDINFO, "game", aBuf);
 	}
 
-	m_aSkins.add(Skin);
+	m_Skins.push_back(Skin);
+	std::sort(m_Skins.begin(), m_Skins.end());
```
Comment (Learath2) [inline src/game/client/components/skins.cpp:299]: Is maybe `std::inplace_merge` or inserting sorted more appropriate here?

File: src/game/localization.cpp (hunk 1 above)
Comment (C0D3D3V) [reply src/game/localization.cpp:44]: I think you placed `CString(` wrong there?! @Learath2

**Comment by Jupeyy** [comment]
> I started commenting on some but I think most of your `push_back`s are good candidates to be `emplace_back`. Otherwise, I think this looks good.

Probs wont make any difference here, bcs the object is temporary created in a stack var anyway (not a pure r-value), does it?

File: src/game/client/components/skins.cpp (hunk 2 above)
Comment (Robyt3) [reply src/game/client/components/skins.cpp]: Done. Had to add `const` to make it compile.

File: src/game/client/components/skins.cpp (hunk 3 above)
Comment (Robyt3) [reply src/game/client/components/skins.cpp]: I replaced all instances of

```
m_Items.push_back(Item);
std::sort(m_Items.begin(), m_Items.end());
```

with

```
m_Items.insert(std::lower_bound(m_Items.begin(), m_Items.end(), Item), Item);
```

File: src/game/editor/editor.cpp [hunk 4]
```diff
@@ -4710,10 +4710,11 @@ void CEditor::FilelistPopulate(int StorageType)
 		Item.m_IsDir = true;
 		Item.m_IsLink = true;
 		Item.m_StorageType = IStorage::TYPE_SAVE;
-		m_FileList.add(Item);
+		m_FileList.push_back(Item);
```
[the remark this answers is not in the document]
Comment (Robyt3) [reply src/game/editor/editor.cpp:4713]: Would require too convoluted constructors here IMO.

File: src/game/editor/editor.cpp [hunk 5]
```diff
@@ -4331,7 +4331,7 @@ static int EditorListdirCallback(const char *pName, int IsDir, int StorageType,
 	Item.m_IsDir = IsDir != 0;
 	Item.m_IsLink = false;
 	Item.m_StorageType = StorageType;
-	pEditor->m_FileList.add(Item);
+	pEditor->m_FileList.push_back(Item);
```
[the remark this answers is not in the document]
Comment (Robyt3) [reply src/game/editor/editor.cpp:4334]: Would require too convoluted constructors here IMO.

**Comment by Robyt3** [comment]
I replaced some `push_back` with `emplace_back`, but I don't want to add too many complicated constructors for every other class in this PR.

**Comment by Learath2** [comment]
> > I started commenting on some but I think most of your `push_back`s are good candidates to be `emplace_back`. Otherwise, I think this looks good.
> 
> Probs wont make any difference here, bcs the object is temporary created in a stack var anyway (not a pure r-value), does it?

The way I miswrote it in that one suggestion, yes. The way it was actually done it shouldn't create any temporary.

File: src/game/client/components/skins.cpp (hunk 3 above)
Comment (Learath2) [reply src/game/client/components/skins.cpp]: Btw, I wasn't completely sure if that's the better approach, that's why I phrased it as a question. Not quite sure how the comparisons and swap counts work out when you do `std::sort` N times as you insert vs figuring out where to insert first.

Comment (Robyt3) [reply src/game/client/components/skins.cpp]: I did some quick benchmarks (times are in ms):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">



<body link="#0563C1" vlink="#954F72">



N | SortOnce | SortEach | Insert |  
-- | -- | -- | -- | --
10 | 371 | 648 | 452 | https://quick-bench.com/q/6pdlu35MhdYA_W4INuzS3Yzvr10
25 | 538 | 2115 | 807 | https://quick-bench.com/q/3oRXXQFBIk-cSOpvHDR9NzOB-hs
50 | 901 | 9554 | 1610 | https://quick-bench.com/q/GVwbe5VYx_uzz0pJ5e8wOBab8fU
100 | 1613 | 55111 | 3403 | https://quick-bench.com/q/RGgY35umHsJOHB7KVvrKB4t3q60
500 | 8653 |   | 22415 | https://quick-bench.com/q/te0NtGHquJU982J41hSaanBcn5s
1000 | 18524 |   | 51348 | https://quick-bench.com/q/te0NtGHquJU982J41hSaanBcn5s



</body>

</html>

grafik

So this roughly shows that sorting only once at the end would be the best. Inserting at the correct position would still be a lot better than sorting the entire vector every time.

**Comment by heinrich5991** [comment]
It seems like we want `v` as a prefix for `std::vector` variables: #5209.

**Comment by Robyt3** [comment]
Can we perform the rename in a separate PR that fixes all names of std containers? I'd like to avoid conflicts with the commits replacing tl/array.

**Comment by heinrich5991** [comment]
The commits already rename the variables as far as I can see.

**Comment by C0D3D3V** [comment]
Looks like he just deleted all prefixes, which makes the names much more attractive ;)
I would still prefer to make the renaming to a uniform prefix automated, it's unbearable manually, always thinking about the prefixes...

**Comment by Robyt3** [comment]
Superseded by #5246.

**Pull request closed without merging**